lib/ukstore/store.c

```diff
@@ -0,0 +1,241 @@
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <uk/store.h>
+#include <errno.h>
+
+
+/* The starting point of all dynamic folders for each library */
+static struct uk_list_head dynamic_heads[__UK_STORE_COUNT] = { NULL, };
+
+/* static struct uk_store_folder *static_folders[2 * __UK_STORE_COUNT] */
+#include <uk/_store_array.h>
+
+/**
+ * Adds a folder to the dynamic list of folders
+ *
+ * @param library_id the library list to add to
+ * @param folder the dynamic folder to be added
+ * @return 0 on success and < 0 on failure
+ */
+int
+_uk_store_add_folder(unsigned int library_id, struct uk_store_folder *folder)
+{
+	if (unlikely(library_id >= __UK_STORE_COUNT))
+		return -EINVAL;
+
+	if (!dynamic_heads[library_id].next)
+		UK_INIT_LIST_HEAD(&dynamic_heads[library_id]);
+
+	uk_list_add(&folder->folder_head, &dynamic_heads[library_id]);
+
+	return 0;
+}
+
+/**
+ * Searches for an folder in a library and returns it.
+ *
+ * @param library_id the library id to search in
+ * @param name the name of the folder to search for
+ * @return the found folder or NULL
+ */
+struct uk_store_folder *
+_uk_store_get_dynamic_folder(unsigned int library_id, const char *name)
+{
+	struct uk_store_folder *res = NULL;
+
+	if (!dynamic_heads[library_id].next)
+		return NULL;
+
+	uk_list_for_each_entry(res, &dynamic_heads[library_id], folder_head)
+		if (!strcmp(res->folder_name, name))
+			return res;
+
+	return NULL;
+}
+
+/**
+ * Searches for an folder in a library and returns it.
+ *
+ * @param library_id the library id to search in
+ * @param name the name of the folder to search for
+ * @return the found folder or NULL
+ */
+struct uk_store_folder *
+_uk_store_get_static_folder(unsigned int library_id, const char *name)
+{
+	struct uk_store_folder *sec_start = static_folders[2 * library_id];
+	struct uk_store_folder *sec_end = static_folders[2 * library_id + 1];
+
+	if (!sec_start)
+		return NULL;
+
+	for (int i = 0; &sec_start[i] != sec_end; ++i)
+		if (!strcmp(sec_start[i].folder_name, name))
+			return &sec_start[i];
+
+	return NULL;
+}
+
+/**
+ * Searches for an entry in a folder and returns it. Increases the refcount.
+ *
+ * @param folder the folder to search in
+ * @param name the name of the entry to search for
+ * @return the found entry or NULL
+ */
+struct uk_store_entry *
+uk_store_get_entry(struct uk_store_folder *folder, const char *name)
+{
+	struct uk_store_folder_entry *res = NULL;
+
+	uk_list_for_each_entry(res, &folder->entry_head, list_head)
+		if (!strcmp(res->entry->entry_name, name))
+			break;
+
+	if (res) {
+		ukarch_inc(&res->refcount.counter);
+		return res->entry;
+	}
+
+	return NULL;
+}
+
+/**
+ * Calls the saved setter with the value in `in`
+ *
+ * @param entry the entry to call from
+ * @param in the value to give to the setter
+ */
+int
+uk_store_set_value(struct uk_store_entry *entry, void *in)
```

Comment:
I was wondering if it is better to provide one function for each setter type, like
``` c
int uk_store_set_u8(struct uk_store_entry *e, __u8 val);
int uk_store_set_u16(struct uk_store_entry *e, __u16 val);
int uk_store_set_u32(struct uk_store_entry *e, __u32 val);
int uk_store_set_charp(struct uk_store_entry *e, const char *val);
```
The function would do the corresponding casting to the target type. Especially when you set a `const char *` to an `__u8` entry, you may want to do the appropriate casting with `atoi()` etc. You could also make sure that a value is within range and return `-ERANGE` otherwise (we should have min/max defined for each type).
You could do a generic switch with a wrapper macro in the header, like:

``` c
#define uk_store_set_value(entry, intype, in) \
   uk_store_set_value_ ## intype ((entry), (in))
```

Reply:
This seems fine. Though, I don't really get why we should check if values are in a range.  If someone calls a `__u8` function with an `__s32` that's on them, and they should take care to not lose information when casting.

Reply:
Also, what purpose does `enum uk_store_entry_type type;` have if you give the type to it like `#define uk_store_set_value(entry, intype, in)`. Or did you suggest this macro for ease of use?

Reply:
To sum it up: the current `uk_store_set_value` function will do the switch based on the type. Afterwards it uses your suggested macro to call the specific function based on the type and return its value.

Did I get this right? I am unsure what will happen to the current `uk_store_set_value` function.